errors/error_test.go

```diff
@@ -0,0 +1,217 @@
+	"github.com/mineiros-io/terramate/errors"
+)
+
+var E = errors.E
+
+var (
+	syntaxError   errors.Kind = "syntax error"
+	tmSchemaError errors.Kind = "terramate schema error"
+	tfSchemaError errors.Kind = "terraform schema error"
+)
+
+func TestNoArgs(t *testing.T) {
+	defer func() {
+		err := recover()
+		if err == nil {
+			t.Fatal("E() did not panic")
+		}
+	}()
+	_ = E()
+}
+
+func TestEmptyTopLevelError(t *testing.T) {
+	defer func() {
+		err := recover()
+		if err == nil {
+			t.Fatal("E() did not panic")
+		}
+	}()
+	_ = E(stderrors.New("test"))
+}
+
+func TestUnknownTypesWithNoFormat(t *testing.T) {
+	defer func() {
+		err := recover()
+		if err == nil {
+			t.Fatal("E() did not panic")
+		}
+	}()
+	_ = E(10, true, 2.5)
+}
+
+func TestErrorString(t *testing.T) {
+	type testcase struct {
+		name string
+		err  error
+		want string
+	}
+
+	for _, tc := range []testcase{
+		{
+			name: "simple message",
+			err:  E("error"),
+			want: "error",
+		},
+		{
+			name: "simple formatted message",
+			err:  E("fmted %s %t %d%d", "string", true, 13, 37),
+			want: "fmted string true 1337",
+		},
+		{
+			name: "error aware types are not use in the format",
+			err:  E("fmted %s %t %d%d", "string", errors.Kind("test"), true, 13, 37),
+			want: "test: fmted string true 1337",
+		},
+		{
+			name: "all non recognized types are format args",
+			err:  E(errors.Kind("test"), true, 13, 37, "fmted %t %d%d"),
+			want: "test: fmted true 1337",
+		},
+		{
+			name: "simple message with kind",
+			err:  E(syntaxError, "failed to parse config"),
+			want: fmt("%s: failed to parse config", syntaxError),
+		},
+		{
+			name: "the kind of previous error is promoted if new one lacks it",
+			err:  E("failed to parse config", E(syntaxError, "unexpected IDENT")),
+			want: fmt("%s: failed to parse config: unexpected IDENT", syntaxError),
+		},
+		{
+			name: "multiple different error kinds",
+			err: E(
+				tmSchemaError, "failed to parse config",
+				E(tfSchemaError, "malformed terraform code"),
+			),
+			want: fmt(
+				"%s: failed to parse config: %s: malformed terraform code",
+				tmSchemaError, tfSchemaError,
+			),
+		},
+		{
+			name: "the file range gets promoted if current error lacks the file context",
+			err: E("failed to parse config",
+				E(tmSchemaError, hcl.Range{
+					Filename: "test.tm",
+					Start:    hcl.Pos{Line: 1, Column: 5, Byte: 3},
+					End:      hcl.Pos{Line: 1, Column: 10, Byte: 13},
+				}, "unexpected attribute name"),
+			),
+			want: fmt("test.tm:1,5-10: %s: failed to parse config: unexpected attribute name",
+				tmSchemaError),
+		},
+		{
+			name: "nested errors",
+			err:  E("1", E("2", E("3"))),
+			want: "1: 2: 3",
+		},
+		{
+			name: "nested errors with last one with range",
+			err: E("1", E("2", E("3", hcl.Range{
+				Filename: "test.tm",
+				Start:    hcl.Pos{Line: 1, Column: 5, Byte: 3},
+				End:      hcl.Pos{Line: 1, Column: 10, Byte: 13},
+			})),
+			),
+			want: "test.tm:1,5-10: 1: 2: 3",
+		},
```

Comment:
I think we are missing the behavior when building errors using hcl.Diagnostics here.

Reply:
fixed